### DISCUSSION TITLE
Trouble with esp32-s3fn8 chip

### DISCUSSION THREAD

**Discussion by theidealist** [micropython/micropython#12332] [ESP32] [answered]
Hello, I recently bought this [Heltec HiLetgo ESP32 Built-in OLED WiFi Kit 32](https://www.amazon.com/dp/B07DKD79Y9?psc=1&ref=ppx_yo2ov_dt_b_product_details) module from Amazon.  I had bought one a couple of years back and gotten micropython setup and running on it no problem and kinda loved it for a home automation project.

However, there's been a hardware revision on this that I didn't notice before purchasing again.  It's now an esp32-s3fn8 chip.  I cannot seem to find the magic firmware + esptool.py invocation to get it to boot properly.  The closest I've gotten is the v1.18 and v1.18 .bin files [here](https://micropython.org/download/ESP32_GENERIC_S3/).

I'm more of a software guy than a micro-controller guy, so I don't really understand the difference between PSRAM SPIRAM, or "Octal-SPIRAM".  The spec on amazon lists this as having "384KB ROM; 512KB SRAM; 16KB RTC SRAM; 8MB SiP Flash".

Right now, with v1.18 or v.19 .bin files loaded at `-z 0x0` I get this:
```
�ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x420daf32
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fcd0108,len:0xf24
load:0x403b6000,len:0xc2c
load:0x403ba000,len:0x2d60
entry 0x403b61f4
E (224) psram: PSRAM ID read error: 0x00ffffff
E (224) spiram: SPI RAM enabled but initialization failed. Bailing out.
E (10) spiram: SPI RAM not initialized

abort() was called at PC 0x42005f98 on core 1

Backtrace:0x40375ba6:0x3fce4b900x403811ad:0x3fce4bb0 0x403874c6:0x3fce4bd0 0x42005f98:0x3fce4c40 0x4200675f:0x3fce4c60 
ELF file SHA256: 8fd2069ebbdfbfa0
Rebooting...
```
Any ideas?  Do I need to build micropython from source?

**Comment by shariltumin** [comment, marked as answer]
ESP32-S3FN8 8 MB (Quad SPI) Flash without SPIRAM. I would try [v1.20.0-398](https://micropython.org/resources/firmware/ESP32_GENERIC_S3-20230830-unstable-v1.20.0-398-gb2b5bcce2.bin).

It automatically detects and works with or without SPIRAM.

Give it a try, it might work on your board.

**Comment by theidealist** [comment]
That appears to have totally worked!  I don't know why it didn't occur to me to try a nightly build :facepalm: 

Many, many thanks!

**Comment by theidealist** [comment]
So, the nightly build boots and the REPL works, but the pinout seems unrelated to the documentation for my board.  Is that normal?  
The pinout diagram with my board looks like:
![image](https://github.com/micropython/micropython/assets/1173291/70d146f5-f38e-49ec-b7fc-2752deabc262)
And the circuit diagram looks like:
![image](https://github.com/micropython/micropython/assets/1173291/1a01bdf6-be04-47e2-aeea-2e93564789d9)

But the micropython firmware calls any pin above 21 invalid.  What am I missing?

**Comment by robert-hh** [comment]
I just tried with a ESP32S3, and I can use GPIO > 21. Only not all numbers are valid. The next beyond 21 is 33 for non_octal_ram devices, or 38 - 46 for all of them.

**Comment by shariltumin** [comment]
Below are the pin definitions defined in "micropython/ports/esp32/machine_pin.h". (only the part for ESP32S3 is shown).

```
...
#elif CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3

#define MICROPY_HW_ENABLE_GPIO0 (1)
#define MICROPY_HW_ENABLE_GPIO1 (1)
#define MICROPY_HW_ENABLE_GPIO2 (1)
#define MICROPY_HW_ENABLE_GPIO3 (1)
#define MICROPY_HW_ENABLE_GPIO4 (1)
#define MICROPY_HW_ENABLE_GPIO5 (1)
#define MICROPY_HW_ENABLE_GPIO6 (1)
#define MICROPY_HW_ENABLE_GPIO7 (1)
#define MICROPY_HW_ENABLE_GPIO8 (1)
#define MICROPY_HW_ENABLE_GPIO9 (1)
#define MICROPY_HW_ENABLE_GPIO10 (1)
#define MICROPY_HW_ENABLE_GPIO11 (1)
#define MICROPY_HW_ENABLE_GPIO12 (1)
#define MICROPY_HW_ENABLE_GPIO13 (1)
#define MICROPY_HW_ENABLE_GPIO14 (1)
#define MICROPY_HW_ENABLE_GPIO15 (1)
#define MICROPY_HW_ENABLE_GPIO16 (1)
#define MICROPY_HW_ENABLE_GPIO17 (1)
#define MICROPY_HW_ENABLE_GPIO18 (1)
#if !CONFIG_USB_CDC_ENABLED
#define MICROPY_HW_ENABLE_GPIO19 (1)
#define MICROPY_HW_ENABLE_GPIO20 (1)
#endif
#define MICROPY_HW_ENABLE_GPIO21 (1)
#if !CONFIG_SPIRAM
#define MICROPY_HW_ENABLE_GPIO26 (1)
#endif
#if !CONFIG_SPIRAM_MODE_OCT
#define MICROPY_HW_ENABLE_GPIO33 (1)
#define MICROPY_HW_ENABLE_GPIO34 (1)
#define MICROPY_HW_ENABLE_GPIO35 (1)
#define MICROPY_HW_ENABLE_GPIO36 (1)
#define MICROPY_HW_ENABLE_GPIO37 (1)
#endif
#define MICROPY_HW_ENABLE_GPIO38 (1)
#define MICROPY_HW_ENABLE_GPIO39 (1)
#define MICROPY_HW_ENABLE_GPIO40 (1)
#define MICROPY_HW_ENABLE_GPIO41 (1)
#define MICROPY_HW_ENABLE_GPIO42 (1)
#define MICROPY_HW_ENABLE_GPIO43 (1)
#define MICROPY_HW_ENABLE_GPIO44 (1)
#define MICROPY_HW_ENABLE_GPIO45 (1)
#define MICROPY_HW_ENABLE_GPIO46 (1)
#if CONFIG_IDF_TARGET_ESP32S3 && MICROPY_HW_ESP32S3_EXTENDED_IO
#define MICROPY_HW_ENABLE_GPIO47 (1)
#define MICROPY_HW_ENABLE_GPIO48 (1)
#endif
...

```

To get to the pins in the firmware as shown in your board's pinout diagram, you will probably need to build your own custom firmware. You can start by copying "ESP32_GENERIC_S3" and deleting all spiram/SPIRAM references in "mpconfigboard.cmake".

**Reply by theidealist** [reply]
Okay, so I finally managed to do as you suggested.  I setup the esp-idf environment, ran through and built the `ESP32_GENERIC_S3` firmware, uploaded it to the board and it emits psram errors like this on startup:
```E (277) quad_psram: PSRAM ID read error: 0x00ffffff, PSRAM chip not found or not supported, or wrong PSRAM line mode
E (277) esp_psram: PSRAM enabled but initialization failed. Bailing out.
```
I copied the `ESP32_GENERIC_S3` to a new directory.  Then I removed the reference to `boards/sdkconfig.spiram_sx` in the new directory's `mpconfigboard.cmake`.  It looks like this now:
```
set(SDKCONFIG_DEFAULTS
    boards/sdkconfig.base
    boards/sdkconfig.usb
    boards/sdkconfig.ble
#    boards/sdkconfig.spiram_sx
    boards/ESP32_HELTEC_S3/sdkconfig.board
)
```
Then I built and deployed that firmware (`make BOARD=NEW_DIRECTORY`) and it seems to have worked!  No psram errors anymore.  But it still doesn't know about the pins.  Is there documentation somewhere that I could follow to further customize this board so that it matches the schematic and pin layout above?  I don't know much how to do that...

I get the following now on startup:
```ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3810,len:0xf58
load:0x403c9700,len:0xb38
load:0x403cc700,len:0x2c80
entry 0x403c989c
Traceback (most recent call last):
  File "main.py", line 34, in <module>
ValueError: invalid pin
MicroPython v1.20.0-427-gf8bd6778c-dirty on 2023-09-03; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
```

Thanks again for any and all help!

**Reply by shariltumin** [reply]
What is on line 34 of your main.py? What pin is the statement referring to?

**Reply by theidealist** [reply]
```
# On heltec dev board the WHITE led is connected to pin 25
WHITE_LED = machine.Pin(35, machine.Pin.OUT)
WHITE_LED.value(0)
```
Just trying to see if I can map the pins in the GPIO diagram to anything meaningful at this point.  The reason the comment doesn't match the code is because I've been trying out different pins.  It was pin 25 when I bought the earlier revision of this board ~3 years ago.

**Reply by shariltumin** [reply]
What do you have in your boards/ESP32_HELTEC_S3/sdkconfig.board?
Do you have CONFIG_SPIRAM_MODE_OCT set? You might want to delete that.

**Reply by theidealist** [reply]
It's the same as what's in `boards/ESP32_GENERIC_S3` -- I haven't touched anything else.  This is what it looks like:
```
CONFIG_ESPTOOLPY_FLASHMODE_QIO=y
CONFIG_ESPTOOLPY_FLASHFREQ_80M=y
CONFIG_ESPTOOLPY_AFTER_NORESET=y

CONFIG_ESPTOOLPY_FLASHSIZE_4MB=
CONFIG_ESPTOOLPY_FLASHSIZE_8MB=y
CONFIG_ESPTOOLPY_FLASHSIZE_16MB=
CONFIG_PARTITION_TABLE_CUSTOM=y
CONFIG_PARTITION_TABLE_CUSTOM_FILENAME="partitions-8MiB.csv"
```

I removed the mention of `SPIRAM` in `mpconfigboard.cmake` and rebuilt.  Reflashed and reset, but same error.

I'm just not sure how to use the pin diagram above to make this custom firmware match up...

Thanks for the help!

**Reply by shariltumin** [reply]
Can you show me your mpconfigboard.cmake file?

**Reply by theidealist** [reply]
Sure!
```
set(IDF_TARGET esp32s3)

set(SDKCONFIG_DEFAULTS
    boards/sdkconfig.base
    boards/sdkconfig.usb
    boards/sdkconfig.ble
#    boards/sdkconfig.spiram_sx
    boards/ESP32_HELTEC_S3/sdkconfig.board
)
```

**Reply by shariltumin** [reply]
I dont see any problem there. I'm running out idea, but there is one thing you can do. Edit your machine_pin.h file.

Replace these:

```
#if !CONFIG_SPIRAM
#define MICROPY_HW_ENABLE_GPIO26 (1)
#endif
#if !CONFIG_SPIRAM_MODE_OCT
#define MICROPY_HW_ENABLE_GPIO33 (1)
#define MICROPY_HW_ENABLE_GPIO34 (1)
#define MICROPY_HW_ENABLE_GPIO35 (1)
#define MICROPY_HW_ENABLE_GPIO36 (1)
#define MICROPY_HW_ENABLE_GPIO37 (1)
#endif
```

With these:
```
#define MICROPY_HW_ENABLE_GPIO26 (1)
#define MICROPY_HW_ENABLE_GPIO33 (1)
#define MICROPY_HW_ENABLE_GPIO34 (1)
#define MICROPY_HW_ENABLE_GPIO35 (1)
#define MICROPY_HW_ENABLE_GPIO36 (1)
#define MICROPY_HW_ENABLE_GPIO37 (1)
```

And replace these:
```
#if CONFIG_IDF_TARGET_ESP32S3 && MICROPY_HW_ESP32S3_EXTENDED_IO
#define MICROPY_HW_ENABLE_GPIO47 (1)
#define MICROPY_HW_ENABLE_GPIO48 (1)
#endif
```

With these:
```
#define MICROPY_HW_ENABLE_GPIO47 (1)
#define MICROPY_HW_ENABLE_GPIO48 (1)
```

Rebuild and reflash your firmware. You should have defined all the higher GPIO26 and GPIO33 to GPIO48 plus the ones you have from before.

Open REPL and try:
```
from machine import Pin
p = Pin(26, Pin.OUT)
p = Pin(33, Pin.OUT)
...
p = Pin(48, Pin.OUT)
```

If one of the GPIO is not defined in the firmware, you will get an error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid pin
```

I hope this is going to work.

**Reply by theidealist** [reply]
Works!!!  At least for the LED!  Thank you so much for your time and helpfulness @shariltumin!  Hopefully I can get the OLED working soon.

I will try not to pester you anymore!

:bowing_man:

**Reply by plonxyz** [reply]
Hi , just found this issue. I'm trying to flash the current micropython release on the heltec LoRa32 V3 , ( ESP32-S3FN8 ). Seems it's stuck in a loop . could anyone please provide me the [v1.20.0-398] firmware? seems that one worked in the past , thanks!

**Reply by theidealist** [reply]
I found it in my downloads and put it up [here](https://drive.google.com/file/d/1gvIVWD8x1KxbA7xnsE7Zvs667ys3ffOp/view?usp=sharing).

Did you try the nightly builds?  I'd be surprised if v1.20.0-398 works, but at least one of the current nightly does not.  Surprised, but I suppose not _too_ surprised as the nightly is unstable!  You can get to the nightly builds here:
https://micropython.org/download/ESP32_GENERIC_S3/

Good luck!

**Reply by plonxyz** [reply]
thank you very much @theidealist ! Now my board works , unfortunately all the builds which are available don't work , it gets either stuck in a reboot loop or it crashes while booting

**Reply by theidealist** [reply]
Glad I could help!
